### ISSUE TITLE
Cancelling ReadWriteRef.read can result in a locked-up state

### ISSUE THREAD

**Issue by sirmax** [evolution-gaming/cats-helper#39]
Initially we observed this issue with `FeatureToggled`. It went like so:
- we have a feature-toggled resource, toggled on and under some use
- we toggle it off
- later we toggle it on, but it does not go back.
Heap dumps showed that an internal `rwLock: ReadWriteRef` had a positive number of active tasks despite them being complete long time ago.

We've narrowed down the problem to the incorrect cancellation handling in `ReadWriteRef`, specifically this bit
```scala
val read: Resource[F, A] = {
  Resource.suspend {
    Concurrent.cancelableF[F, Resource[F, A]] { cb =>
      …
      stateRef
        .modify { // acquisition
            …
            val effect = … // emit a Resource that modifies stateRef on release
            val cancel = F.unit
            s1 -> (effect as cancel)
        }
        .flatten
    }
  }
```
In a nutshell, we have two `stateRef` updates here: one for "acquisition" and one for "release", but the problem is that they happen on different "layers", and its quite possible that only "acquisition" happens but not "release" when the `read.use(…)` gets cancelled.
